File: packages/companion/swift/Sources/BridgethingCompanion/DeviceLogRing.swift
import Foundation

public struct DeviceLogRecord: Sendable {
  public let seq: UInt64
  public let timestampMs: Double
  public let level: String
  public let message: String
}

/// Process-wide ring of device-log lines
public final class DeviceLogRing: @unchecked Sendable {
  public static let shared = DeviceLogRing()

  private let limit: Int
  private let lock = NSLock()
  private var records: [DeviceLogRecord] = []
  private var seqCounter: UInt64 = 0

  public init(limit: Int = 2000) {
    self.limit = limit
  }

  public func push(level: String, message: String) {
    lock.lock()
    seqCounter += 1
    records.append(
      DeviceLogRecord(
        seq: seqCounter,
        timestampMs: Date().timeIntervalSince1970 * 1000,
        level: level,
        message: message
      )
    )
    if records.count > limit { records.removeFirst(records.count - limit) }
    lock.unlock()
  }

  public func tail(limit: Int) -> [DeviceLogRecord] {
    lock.lock(); defer { lock.unlock() }
    guard limit < records.count else { return records }
    return Array(records.suffix(limit))
  }
}

File: packages/companion/kotlin/companion/src/main/kotlin/com/bridgething/companion/DeviceLogRing.kt
package com.bridgething.companion

public data class DeviceLogRecord(
    val seq: Long,
    val timestampMs: Double,
    val level: String,
    val message: String,
)

/**
 * Process-wide ring of device-log lines (daemon tracing forwarded over the gateway,
 * plus companion-origin lines). The live `logObserver` path carries entries in
 * real-time while foregrounded; this durable tail lets the RN side backfill the gap
 * opened while the app is backgrounded and the JS thread is suspended.
 */
public object DeviceLogRing {
    private const val LIMIT = 2000

    private val lock = Any()
    private val records = ArrayDeque<DeviceLogRecord>()
    private var seqCounter = 0L

    public fun push(level: String, message: String) {
        synchronized(lock) {
            seqCounter += 1
            records.addLast(DeviceLogRecord(seqCounter, System.currentTimeMillis().toDouble(), level, message))
            while (records.size > LIMIT) records.removeFirst()
        }
    }

    public fun tail(limit: Int): List<DeviceLogRecord> = synchronized(lock) {
        if (limit >= records.size) records.toList() else records.toList().takeLast(limit)
    }
}
